File: src/main/java/kitchenpos/menus/tobe/domain/Menu.java
package kitchenpos.menus.tobe.domain;

import jakarta.persistence.*;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

@Table(name = "menu")
@Entity(name = "tobeMenu")
public class Menu {
    @Column(name = "id", columnDefinition = "binary(16)")
    @Id
    private UUID id;

    @Embedded
    private MenuName name;

    @Embedded
    private MenuPrice price;

    @ManyToOne(optional = false)
    @JoinColumn(
            name = "menu_group_id",
            columnDefinition = "binary(16)",
            foreignKey = @ForeignKey(name = "fk_menu_to_menu_group")
    )
    private MenuGroup menuGroup;

    @Column(name = "displayed", nullable = false)
    private boolean displayed;

    @Embedded
    private MenuProducts menuProducts;

    @Transient
    private UUID menuGroupId;

    protected Menu() {
    }

    public Menu(UUID id, MenuName name, MenuPrice price, MenuGroup menuGroup, boolean displayed, MenuProductCatalog menuProductCatalog) {
        validatePrice(price, menuProductCatalog.calculateTotalPrice());
        this.id = id;
        this.name = name;
        this.price = price;
        this.menuGroup = menuGroup;
        this.displayed = displayed;
        this.menuProducts = new MenuProducts(menuProductCatalog);
    }

    private void validatePrice(MenuPrice price, BigDecimal total) {
        if (!price.isGreaterThan(total)) {
            throw new IllegalArgumentException("메뉴 가격이 전체 메뉴 상품의 가격 합보다 같거나 작아야 한다");
        }
    }

    public UUID getId() {
        return id;
    }

    public void changePrice(MenuPrice price, ProductInfos productInfos) {
        validatePrice(price, menuProducts.calculateTotalPrice(productInfos));
        this.price = price;
    }

    public List<UUID> findProductIds() {
        return menuProducts.getProductIds();
    }

    public void display(ProductInfos productInfos) {
        validatePrice(price, menuProducts.calculateTotalPrice(productInfos));
        this.displayed = true;
    }

    public void hide() {
        this.displayed = false;
    }

    public MenuPrice getPrice() {
        return price;
    }
}

File: src/main/java/kitchenpos/menus/tobe/domain/Menu.kt
package kitchenpos.menus.tobe.domain

import jakarta.persistence.*
import kitchenpos.common.Price
import java.util.*

@Table(name = "menu")
@Entity
class Menu private constructor(
    menuGroup: MenuGroup,
    name: MenuName,
    price: Price,
    displayStatus: Boolean,
    menuProducts: MenuProducts
) {
    @Column(name = "id", columnDefinition = "binary(16)")
    @Id
    val id: UUID = UUID.randomUUID()

    @ManyToOne(optional = false)
    @JoinColumn(
        name = "menu_group_id",
        columnDefinition = "binary(16)",
        foreignKey = ForeignKey(name = "fk_menu_to_menu_group")
    )
    val menuGroup: MenuGroup = menuGroup

    @Column(name = "name", nullable = false)
    val name: MenuName = name

    @Column(name = "price", nullable = false)
    var price: Price = price
        private set

    @Column(name = "displayed", nullable = false)
    var displayStatus: Boolean = displayStatus
        private set

    @Embedded
    val menuProducts: MenuProducts = menuProducts

    private val isMenuPriceValid: Boolean
        get() = price <= menuProducts.totalPrice

    companion object {
        fun of(
            menuGroup: MenuGroup,
            name: MenuName,
            price: Price,
            displayStatus: Boolean,
            menuProducts: MenuProducts,
        ): Menu = Menu(
            menuGroup = menuGroup,
            name = name,
            price = price,
            displayStatus = displayStatus,
            menuProducts = menuProducts
        )
    }

    init {
        validateMenuPrice()
        this.menuProducts.apply(this)
    }

    fun activateDisplayStatus() {
        this.displayStatus = true

        validateMenuPrice()
    }

    fun inActivateDisplayStatus() {
        this.displayStatus = false
    }

    fun changePrice(price: Price) {
        this.price = price

        validateMenuPrice()
    }

    fun changeMenuProduct(productId: UUID, productPrice: Price) {
        menuProducts.changeMenuProduct(productId, productPrice)

        if (!isMenuPriceValid) {
            inActivateDisplayStatus()
        }
    }

    private fun validateMenuPrice() {
        if (displayStatus && !isMenuPriceValid) {
            throw IllegalArgumentException("메뉴의 가격이 메뉴상품 가격의 총합보다 클 수 없습니다")
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Menu

        return id == other.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

}
